kopf/reactor/causation.py

```diff
@@ -0,0 +1,132 @@
+"""
+Detection of the event causes, based on the resource state.
+
+The low-level watch-events are highly limited in information on what
+caused them, and they only notify that the object was changed somehow:
+
+* ``ADDED`` for the newly created objects (or for the first-time listing).
+* ``MODIFIED`` for the changes of any field, be that metainfo, spec, or status.
+* ``DELETED`` for the actual deletion of the object post-factum.
+
+The conversion of the low-level *events* to the high-level *causes* is done by
+checking the object's state and comparing it to the saved last-seen state.
+
+This allows to track which specific fields were changed, and are those changes
```

Comment:
"and if those changes were important enough to call the handlers..." unless I misinterpreted it 🤔